### ISSUE TITLE
Mobilenet 2 bits weight and activation

### ISSUE THREAD

**Issue by giop98** [Xilinx/brevitas#430]
Hi,

thanks for your amazing work.
I'm trying to implement a quantized version, with 2 bits weights and 2 bits activation, of mobilenet. I would like to execute the whole flow, so starting from the model, training it and then deploying it to hardware with FINN.

I started from the model that is in this repository (https://github.com/Xilinx/brevitas/blob/master/src/brevitas_examples/imagenet_classification/models/mobilenetv1.py). 
I kept the first layer quantization to 8 bits and I tried to change all the other layer quantization to 2 bits. 
I then implemented the training loop for imagenet, by following the one in the PyTorch repository (https://github.com/pytorch/examples/tree/main/imagenet).
However, during training, after few hundreds batches of the first epoch (I'm training with 128 batch size), the model starts showing a loss equal to NaN. 
As a further investigation I also tried to run two different versions with different quantization, in order to understand what is causing the problem:

- 2 bits weight and 4 bits activation: This model, after 3 epochs, return NaN
- 4 bits weight and 2 bits activation: This model is still running but it seems to perform better and after 3 epochs is still learning without any issue so far

As a conclusion it seems that the problem is related to the low number of bits for the weight. Is there any working implementation of mobilenet 2bits 2bits? If not, is there a way to fix the loss going to NaN in the 2bits 2bits  implementation?

Thanks for your time

**Comment by volcacius** [comment]
Hey Giovanni,

Try with the dev branch, when training goes NaN usually it's because some scale factors go to zero. You can avoid that by setting a hard lower bound on them with scaling_min_val=, off the top of my head I think I'm now setting it by default in the dev branch to 1e-10 (used to be None). 
In general training 2b depthwise separable weights at ImageNet scale is tricky. I would keep those at a higher precision. If you are especially committed to 2b only you could implement this scheme https://arxiv.org/abs/2105.08952.
